services/graph/pkg/service/v0/tags.go

```diff
@@ -0,0 +1,226 @@
+		w.WriteHeader(http.StatusForbidden)
+		return
+	}
+
+	var currentTags string
+	if m := sres.GetInfo().GetArbitraryMetadata().GetMetadata(); m != nil {
+		currentTags = m["tags"]
+	}
+
+	allTags := revaTags.FromList(currentTags)
+	newTags := strings.Join(assignment.Tags, ",")
+	if !allTags.AddList(newTags) {
+		errorcode.InvalidRequest.Render(w, r, http.StatusBadRequest, "no new tags in createtagsrequest or maximum reached")
+		return
+	}
+
+	resp, err := g.gatewayClient.SetArbitraryMetadata(ctx, &provider.SetArbitraryMetadataRequest{
+		Ref: &provider.Reference{ResourceId: &rid},
+		ArbitraryMetadata: &provider.ArbitraryMetadata{
+			Metadata: map[string]string{
+				"tags": allTags.AsList(),
+			},
+		},
+	})
+	if err != nil || resp.GetStatus().GetCode() != rpc.Code_CODE_OK {
+		g.logger.Error().Err(err).Msg("error setting tags")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	if g.eventsPublisher != nil {
+		ev := events.TagsAdded{
+			Tags: newTags,
+			Ref: &provider.Reference{
+				ResourceId: &rid,
+				Path:       ".",
+			},
+			SpaceOwner: sres.Info.Owner,
+			Executant:  revaCtx.ContextMustGetUser(r.Context()).Id,
+		}
+		if err := events.Publish(g.eventsPublisher, ev); err != nil {
+			g.logger.Error().Err(err).Msg("Failed to publish TagsAdded event")
+		}
+	}
+}
+
+func (g Graph) UnassignTags(w http.ResponseWriter, r *http.Request) {
+	var (
+		unassignment libregraph.TagUnassignment
+		ctx          = r.Context()
+	)
+
+	if err := json.NewDecoder(r.Body).Decode(&unassignment); err != nil {
+		g.logger.Debug().Err(err).Interface("body", r.Body).Msg("could not decode tag assignment request")
+		errorcode.InvalidRequest.Render(w, r, http.StatusBadRequest, "invalid body schema definition")
+		return
+	}
+
+	rid, err := storagespace.ParseID(unassignment.ResourceId)
+	if err != nil {
+		g.logger.Debug().Err(err).Msg("could not parse resourceId")
+		errorcode.InvalidRequest.Render(w, r, http.StatusBadRequest, "invalid resourceId")
+		return
+	}
+
+	sres, err := g.gatewayClient.Stat(ctx, &provider.StatRequest{
+		Ref: &provider.Reference{ResourceId: &rid},
+	})
+	if err != nil {
+		g.logger.Error().Err(err).Msg("error stating file")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	if sres.GetStatus().GetCode() != rpc.Code_CODE_OK {
+		errorcode.InvalidRequest.Render(w, r, http.StatusBadRequest, "can't stat resource")
+		return
+	}
+
+	pm := sres.GetInfo().GetPermissionSet()
+	if pm == nil {
+		g.logger.Error().Err(err).Msg("no permissionset on file")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// it says we need "write access" to set tags. One of those should do
+	if !pm.InitiateFileUpload && !pm.CreateContainer {
+		g.logger.Info().Msg("no permission to create a tag")
+		w.WriteHeader(http.StatusForbidden)
+		return
+	}
+
+	var currentTags string
+	if m := sres.GetInfo().GetArbitraryMetadata().GetMetadata(); m != nil {
+		currentTags = m["tags"]
+	}
```

Comment:
Afaics this code part is 1:1 the same as in `AssignTags`. Can we put this in a func that both are using?